### DISCUSSION TITLE
Command line export/import

### DISCUSSION THREAD

**Discussion by metasean** [rxhanson/Rectangle#803] [Ideas]
(A) I 💟 Rectangle!

(B) I _just_ learned about the [cli only configs](https://github.com/rxhanson/Rectangle/blob/master/TerminalCommands.md) —some of which I've wanted to be able to do for a while— and now I've found out just how to do it.

(C)
For reasons, I've had to set-up new machines for myself on a regular basis at work.  I've finally gotten frustrated enough that I'm starting to automate the process (combo of shell script, ansible config, and brew installs from the ansible config).  

Someone was kind enough to already create a [brew formula for rectangle](https://formulae.brew.sh/cask/rectangle#default).  But I'd also like to be able to automate the process of maintaining my set-up process (i.e. to export my Rectangle config from a script) as well as automating the set-up itself (i.e. importing my Rectangle config from a script.). 

Y'all have already done a great job just by having the export and import options!!!  
Are there command line ways to run them?
If not, can y'all add them? 🙏

**Comment by rxhanson** [comment]
Thanks for the feedback. 

The easiest way is to just copy over the NSUserDefaults, which is where the preferences are persisted. The import/export in Rectangle ultimately is just reading/writing those anyway. See the Preferences Storage section in the readme: https://github.com/rxhanson/Rectangle#preferences-storage

**Reply by metasean** [reply]
Thank you for such an awesome and useful utility and for such a quick response.

The approach you describe was my fallback, so I should be good to go.

Again, Thank you!

**Reply by cadriel** [reply]
I'd like to chime in here and mention that yeah, copying NSUserDefaults works - but having a specific command to import settings (or read a `dotfile` default) is preferable.

I use my own custom `dotfiles` and having to setup scripts to manage this copy is a bit hacky - not to mention if the format of these change over versions then it becomes more of a chore to maintain.

So 👍🏼 to either having a specific command to handle this (and handle differences over time if prefs change..) - or to read from a config file.

**Reply by rxhanson** [reply]
@cadriel good to know. Scripting the app would return the best results, but it's oddly difficult to implement, and would require a decent amount of time to get right. It might be worth it to explore other alternatives, like opening the app with a config file as an argument. I don't quite have the time to investigate at the moment, but would definitely welcome a good PR for it.

**Reply by mavam** [reply]
Chiming in here, as I was about to create a discussion. It's pretty common to store user configuration in `~/.config`, e.g., `~/.config/Rectangle/config.json`.

It would be great if Rectangle at least *read* from this location, allowing users that manage their dotfiles to have a unified approach to version controlling them. NeoVim, Kitty, Fish, and many other tools simply respect the [`$XDG_CONFIG_HOME`](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html).

**Reply by rxhanson** [reply]
@mavam Thanks for the feedback. I see no reason to turn down a good PR where Rectangle reads from a directory as mentioned. Seems like it wouldn't be a big deal to look in multiple places as long as the automatic read is in place. However, I will say that it would seem rather odd if it read from a file but did not write back out to it when config changes, since restarting the app would then re-read that config unless there's logic specifically for telling it to not read it again (which would seem more confusing then just writing back out config as it is modified).

**Reply by mavam** [reply]
Glad to know you are open to this behavior. I'm fully with you that a symmetric approach makes most sense, covering reads and writes. Unlike Rectangle, many terminal applications simply don't need the write path, because their config gets only loaded once at startup.

**Reply by nbelles** [reply]
I have almost the same exact same use case as the original poster. After installing via brew, to get the application going, I have a command `open -a /Applications/Rectangle.app` so that it goes ahead and starts running the app. It would be nice if we could do a similar thing but tack on a path to a .json file with the configuration that was exported before. What I think this could look like: `open -a /Applications/Rectanle.app ~/Downloads/RectangleConfig.json` That would make it just that much easier to get up and running with a new install.

**Reply by yoaquim** [reply]
Was this ever implemented?

**Reply by rxhanson** [reply]
It was not. At the moment, I don't have the bandwidth for it, but I can point someone to the right place for starting on it (shouldn't be too hard).
